### PR TITLE
Migrate to ESLint instead of JSHint, JSCS

### LINKED ISSUE #123: JSCS + ES2015
Our current JSCS setup doesn't work well with ES2015 (and thus not with the babel option). If you choose to include babel it fails and reports syntax errors.

Also, if you the `dist` directory is present it lints that directory too.

### LINKED ISSUE #140: JShint shouldn't raise issues when parsing es2015 code
Currently this is failing. That's kind of related to #123, but this issue is probably going to take a while, so let's make sure we're fixing JSHint ASAP.

### PR DESCRIPTION
**Pull request by zckrs**
fix #140
fix #123 

#### Squashed commits
* 3594a70 - Add ESLint files, devDev
* bbbd02d - ESLint local fix
* 76e021d - Add subgenerator eslint
* e79a6f0 - ESlint templates fix
* 039553d - Remove subgenerator JSHint, JSCS && Update README
* 3fe7058 - Fix codestyle after templating gulpfile with babel
* 0713b18 - Use global .eslintrc in generator and template app
* 8c9dce0 - Unable ESLint rule consistent-return
* e50061d - Write JSON instead of template file
* 40da38f - Remove local ESLint files
* bb9ce90 - Update README
* 00ba7b6 - Use destinationPath to write eslintrc
* 15c0c43 - Configure no-use-before-define for function
* acdb817 - Rename eslint option babel to es2015
* 7df79eb - Complete test
* a1f8b84 - Remove fill package.json with eslint devDep
* e4d3490 - gulpfile is not yet write in es2015
* acd5073 - Apply Yeoman style guide

### CODE REVIEW

**Comment by hemanth** [comment]
:+1: LGTM. //cc @SBoudrias

File: generators/app/index.js [hunk 1]
```diff
@@ -90,14 +90,14 @@ module.exports = generators.Base.extend({
             return false;
           }
 
-          var done = this.async();
+          var cb = this.async();
```
Comment (arthurvr) [inline generators/app/index.js]: Why did you rename this? Please keep unrelated changes out of this commit :wink:

Comment (hemanth) [inline generators/app/index.js]: I noticed and ignored it, `done` sounds better, yeah. :8ball:

Comment (zckrs) [inline generators/app/index.js]: https://github.com/yeoman/generator-node/commit/bbbd02d82ad871f5561ae6958dffb79f82ec8d02#diff-def8dd7c5bf7421ace8ec454d89f771cR85

ESLint error:
```
generators/app/index.js
  93:14  error  done is already declared in the upper scope  no-shadow

```

File: test/.eslintrc [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "env": {
+    "mocha": true
+  }
+}
```
Comment (arthurvr) [inline test/.eslintrc]: Meh. Is an extra file in `test/` really worth it just for one option? Just like we did with [JSHint](https://github.com/yeoman/generator-node/blob/b25e50b89894093901116e442f9f2c1f03bb5317/.jshintrc#L10) I prefer this to go in the main `.eslintrc`.

File: generators/gulp/templates/gulpfile.js [hunk 3]
```diff
@@ -66,7 +58,7 @@ gulp.task('test', ['pre-test'], function (cb) {
 
 gulp.task('coveralls', ['test'], function () {
   if (!process.env.CI) {
-    return;
+    return false;
```
Comment (arthurvr) [inline generators/gulp/templates/gulpfile.js]: Is this also fixing to fix an eslint warning? If so, let's edit the config because this is perfectly valid.

File: test/.eslintrc (hunk 2 above)
Comment (zckrs) [inline test/.eslintrc]: I use [ESLint configuration cascading](]http://eslint.org/docs/user-guide/configuring#configuration-cascading-and-hierarchy)

Do you prefer enable mocha env in whole project ? Or define env mocha with inline comment `/*eslint-env  mocha */` in test/index.js

File: generators/gulp/templates/gulpfile.js (hunk 3 above)
Comment (zckrs) [inline generators/gulp/templates/gulpfile.js]: Can you annonce what ESLint configuration do you want?

Comment (arthurvr) [inline generators/gulp/templates/gulpfile.js]: Not sure but it seems like http://eslint.org/docs/rules/consistent-return - I just want `return;` to be valid in cases like this.

Comment (zckrs) [inline generators/gulp/templates/gulpfile.js]: Ok. Can I use the same configuration across generator and template generated ?

File: test/.eslintrc (hunk 2 above)
Comment (arthurvr) [inline test/.eslintrc]: > Do you prefer enable mocha env in whole project ? Or define env mocha with inline comment /*eslint-env mocha */ in test/index.js

I'm not really sure, but I don't really like this `test/.eslintrc` file too. I would opt to just define it in the main `.eslintrc` as it's not that big of a deal. WDYT @hemanth @SBoudrias?

**Comment by arthurvr** [comment]
For this rest this LGTM. Super cool work! Did you confirm that this fixes https://github.com/yeoman/generator-node/issues/123? Thanks!

File: test/.eslintrc (hunk 2 above)
Comment (zckrs) [inline test/.eslintrc]: To compare [ESLint repo](https://github.com/eslint/eslint/tree/master/tests/lib) use a `.eslintrc` in their test folder.

Env mocha will adds all of the Mocha testing global variables. https://github.com/sindresorhus/globals/blob/master/globals.json#L683

**Comment by zckrs** [comment]
About #123, currently ESLint error provide by `$ gulp static`

#### Without ES2015
```
/home/zckrs/github/sampleNode/test/index.js
  5:4  error  sampleNode is defined but never used  no-unused-vars
```
* Can I delete this unused variable ???

#### With ES2015
```
/home/zckrs/github/sampleNode/lib/index.js
  1:1  error  Illegal export declaration

/home/zckrs/github/sampleNode/test/index.js
  1:1  error  Illegal import declaration
```

To fix this error we should define ESLint env es6 + ecmaFeatures module ([details](http://eslint.org/docs/user-guide/configuring#specifying-environments)).
* We can use cascading `.eslintrc`as [mention](https://github.com/yeoman/generator-node/pull/151#discussion_r35198033) :smile: 
* Or define global `.eslintrc`  + desactivate es6 env in `gulpfile.js` :disappointed:

**Comment by zckrs** [comment]
@arthurvr I push new commit where a single file is used to define ESLint configuration https://github.com/yeoman/generator-node/commit/e1a31fce227f5e787ce649e38384029ff9c9c9ed

File: generators/gulp/templates/gulpfile.js (hunk 3 above)
Comment (zckrs) [inline generators/gulp/templates/gulpfile.js]: Done https://github.com/yeoman/generator-node/commit/ab85836734040e9f315d071c3eac2754e089becc

**Comment by zckrs** [comment]
Last ESLint error returned by `gulp static`:

```
/home/zckrs/github/sampleNode/test/index.js
  4:4  error  sampleNode is defined but never used  no-unused-vars
```

**Comment by SBoudrias** [comment]
I'll take time to go through the PR later.

But at first glance:

1. About the `sampleNode` error, that's fine. The test is missing, so it's normal our build is failing.
2. The `.eslintrc` file is a JSON file. Just construct it inside the generator and use `writeJSON()', don't use a template for this.
3. Can we only update the generated project to use eslint. (we'll just run generator-node on generator-node to update the build configuration)

And finally a question, is this actually validating code style?

**Comment by robbbz** [comment]
I was working on a PR for this and also commented that on the ticket. I can't see how you like to do duplicate work, and I worked on this really long already. I had this in my `eslint` branch if you want to see it.

**Comment by zckrs** [comment]
Sorry @robbz i search branche befote starting but not found.

I will update PR. Thank you @SBoudrias

**Comment by SBoudrias** [comment]
@robbbz Actually, for you to get the full context, these changes were mostly intended for generator-generator. So @zckrs didn't knew about the same issue for generator-node, and it would've been hard to know.

So we just asked him to move his branch to generator-node as generator-generator will compose it in the near future.

This is unfortunate, but it's not like there was a lot of chances for anyone of you two to know someone else was working on the same thing.

Still @robbbz, I'd really like you to review this PR and chip in about what you would've done differently or if there's things that are missing. This would really be useful.

**Comment by zckrs** [comment]
@SBoudrias 
* e50061d write JSON programmatically
* 40da38f remove local ESLint files

> And finally a question, is this actually validating code style?

Actually no, but we should configure the root .eslintrc.
Exist many [stylistic rules](http://eslint.org/docs/rules/#stylistic-issues).
What do you want :-) ?

Soon ESLint will release a first stable version [ESLint v1.0.0-rc-1](https://github.com/eslint/eslint/releases/tag/v1.0.0-rc-1)

File: generators/boilerplate/templates/index.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
-'use strict';
 <% if (babel) { -%>
 export default {};
 <% } else { -%>
+'use strict';
```
Comment (SBoudrias) [inline generators/boilerplate/templates/index.js]: Why are we changing the order here?

File: README.md [hunk 5]
```diff
@@ -84,10 +83,9 @@ Here's a list of our supported options:
 - `babel` (Boolean, default true) include or not a `.babelrc` file.
 - `cli` (Boolean, default false) include or not a `lib/cli.js` file.
 - `editorconfig` (Boolean, default true) include or not a `.editorconfig` file.
+- `eslint` (Boolean, default true) include or not a `.eslintrc` file.
```
Comment (SBoudrias) [inline README.md]: I don't see this option defined - and I don't think we should make it optional.

File: generators/eslint/index.js [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+          ecmaFeatures: {
+            modules: true
+          },
+          env: {
+            es6: true
+          }
+        });
+      }
+
+      this.fs.writeJSON('.eslintrc', eslintrc);
```
Comment (SBoudrias) [inline generators/eslint/index.js]: `this.destinationPath('.eslintrc')`

File: generators/eslint/index.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+var extend = require('deep-extend');
+
+module.exports = generators.Base.extend({
+  constructor: function () {
+    generators.Base.apply(this, arguments);
+
+    this.option('babel', {
+      required: false,
+      defaults: false,
+      desc: 'Compile ES2015 using Babel'
```
Comment (SBoudrias) [inline generators/eslint/index.js]: The description doesn't describe what this generator is doing with babel.

This option should be called `es2015`, "Allow es2015 syntax"

File: generators/gulp/index.js [hunk 8]
```diff
@@ -2,6 +2,10 @@
 var extend = require('deep-extend');
 var generators = require('yeoman-generator');
 
+function stringifyArray(arr) {
```
Comment (SBoudrias) [inline generators/gulp/index.js]: Why did you move this function?

File: generators/gulp/templates/gulpfile.js [hunk 9]
```diff
@@ -1,12 +1,13 @@
+<% if (!babel) { -%>
```
Comment (SBoudrias) [inline generators/gulp/templates/gulpfile.js]: We add `'use strict'` everywhere. Babel or not because es6 didn't remove the strict mode.

It's only babel who add it by default, and that's a mistake on their part.

File: test/eslint.js [hunk 10]
```diff
@@ -3,13 +3,13 @@ var path = require('path');
...
   before(function (done) {
-    helpers.run(path.join(__dirname, '../generators/jshint'))
+    helpers.run(path.join(__dirname, '../generators/eslint'))
       .on('end', done);
   });
 
-  it('creates .jshintrc', function () {
-    assert.file('.jshintrc');
+  it('creates .eslintrc', function () {
+    assert.file('.eslintrc');
```
Comment (SBoudrias) [inline test/eslint.js]: We should assert some stuff about the content.

With and without es6.

File: generators/boilerplate/templates/index.js (hunk 4 above)
Comment (zckrs) [inline generators/boilerplate/templates/index.js]: ESLint error in ES2015 with `'use strict';`
```
lib\index.js
  1:0  error  Unnecessary "use strict" directive  strict
```
Refering to the this http://people.mozilla.org/~jorendorff/es6-draft.html#sec-strict-mode-code

>Module code is always strict mode code.

File: generators/gulp/index.js (hunk 8 above)
Comment (zckrs) [inline generators/gulp/index.js]: ESLint error
```
./generators/gulp\/ndex.js
  73:17  error  stringifyArray was used before it was defined  no-use-before-define
  74:27  error  stringifyArray was used before it was defined  no-use-before-define
```

http://eslint.org/docs/rules/no-use-before-define

File: generators/boilerplate/templates/index.js (hunk 4 above)
Comment (SBoudrias) [inline generators/boilerplate/templates/index.js]: Oh I see! Fine then, let's keep it out.

File: generators/gulp/index.js (hunk 8 above)
Comment (SBoudrias) [inline generators/gulp/index.js]: Can we change that rule then?

File: README.md (hunk 5 above)
Comment (zckrs) [inline README.md]: Done bb9ce907f71a7dd46dc8889625efa92ecfdfca8c

File: generators/eslint/index.js (hunk 6 above)
Comment (zckrs) [inline generators/eslint/index.js]: Done 00ba7b6bc83868ef85406cc9ae2587696412b88f

File: generators/gulp/index.js (hunk 8 above)
Comment (zckrs) [inline generators/gulp/index.js]: Done 15c0c4305c60c426d942474e659a5f788d4266c0

File: generators/gulp/templates/gulpfile.js (hunk 9 above)
Comment (zckrs) [inline generators/gulp/templates/gulpfile.js]: Same error https://github.com/yeoman/generator-node/pull/151#discussion_r35397141

gulpfile should be written in ES2015 if babel ?

File: generators/eslint/index.js (hunk 7 above)
Comment (zckrs) [inline generators/eslint/index.js]: Done acdb8172318207d790bc1785d69eabb7ad45f12e

Should I rename babel option in whole generator ?

File: test/eslint.js (hunk 10 above)
Comment (zckrs) [inline test/eslint.js]: Done 7df79eb4723bcfaa590799ccfbdb98d6f68a0ba6

**Comment by zckrs** [comment]
When ready to merge I will squash commits.

File: generators/gulp/templates/gulpfile.js (hunk 9 above)
Comment (SBoudrias) [inline generators/gulp/templates/gulpfile.js]: Actually no, gulpfile shouldn't be es2015 as it's the file building the babel code. We want this one compatible with node without precompilation.

So we can keep `'use strict';` as it's not a module file.

**Comment by SBoudrias** [comment]
For the stylistic rules, we want to follow Google JS styleguide.

File: generators/gulp/templates/gulpfile.js (hunk 9 above)
Comment (zckrs) [inline generators/gulp/templates/gulpfile.js]: Done

**Comment by zckrs** [comment]
With ESLint preset are not define (like JSCS). https://github.com/eslint/eslint/issues/1669

Do you reference this style guide http://google.github.io/styleguide/javascriptguide.xml ?

**Comment by arthurvr** [comment]
> Do you reference this style guide http://google.github.io/styleguide/javascriptguide.xml ?

No, you can find ours here: http://yeoman.io/contributing/style-guide.html. There are some minor differences such as the space after `function`.

**Comment by zckrs** [comment]
@arthurvr Done :-)

**Comment by SBoudrias** [comment]
Alright, I'll give it a try sometime this weekend.

**Comment by SBoudrias** [comment]
@zckrs Hey, can you squash your commits already so I can merge right away if it's looking good?

**Comment by zckrs** [comment]
@SBoudrias Of course. Done

**Comment by SBoudrias** [comment]
Hey, I feel we're missing a lot of style rules we'll want to enforce using eslint. We can handle this as a follow-up ticket. - We'll block releasing eslint until the styleguide is correctly enforced.

**Comment by SBoudrias** [comment]
Oh and, thanks for all this!!

**Comment by arthurvr** [comment]
Super-duper! Nice work :wink:

**Comment by zckrs** [comment]
You are welcome. =]


Tomorrow I will propose more ESLint rules in https://github.com/yeoman/generator-node/issues/153

**Comment by hemanth** [comment]
:+1:  /me pass some :cookie: to @zckrs
